src/main.rs

```diff
@@ -47,125 +97,293 @@ impl LogLevel {
-        return Ok(());
-    }
-    for path in ["/etc/greenboot/green.d/*.sh"] {
-        for entry in glob(path)?.flatten() {
-            log::info!("running green check {}", entry.to_string_lossy());
-            let output = Command::new("bash").arg("-C").arg(entry).output()?;
-            if !output.status.success() {
-                // combine and print stderr/stdout
-                log::warn!("green script failed...");
-            }
+    }
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        Err(errors)
+    }
+}
+
+/// triggers the diagnostics followed by the action on the outcome
+/// this also handles setting the grub variables and system restart
+fn health_check() -> Result<()> {
+    let config = GreenbootConfig::get_config();
+    log::debug!("{config:?}");
+    handle_motd("healthcheck is in progress")?;
+    match run_diagnostics() {
+        Ok(()) => {
+            log::info!("greenboot health-check passed.");
+            run_green().unwrap_or_else(|errors| {
+                log::error!("There is a problem with green script runner");
+                errors.iter().for_each(|e| log::error!("{e}"));
+            });
+            handle_motd("healthcheck passed - status is GREEN")
+                .unwrap_or_else(|e| log::error!("cannot set motd: {}", e.to_string()));
+            handle_boot_status(true)?;
+            Ok(())
+        }
+        Err(e) => {
+            log::error!("Greenboot error: {e}");
+            handle_motd("healthcheck failed - status is RED")
+                .unwrap_or_else(|e| log::error!("cannot set motd: {}", e.to_string()));
+            run_red().unwrap_or_else(|errors| {
+                log::error!("There is a problem with red script runner");
+                errors.iter().for_each(|e| log::error!("{e}"));
+            });
+            handle_boot_status(false)?;
+            set_boot_counter(config.max_reboot)
+                .unwrap_or_else(|e| log::error!("cannot set boot_counter: {}", e.to_string()));
+            handle_reboot(false)
+                .unwrap_or_else(|e| log::error!("cannot reboot: {}", e.to_string()));
+            Err(e)
         }
     }
-    Ok(())
 }
 
-fn reboot() -> Result<(), Error> {
-    Command::new("systemctl").arg("reboot").spawn()?;
-    Ok(())
+/// initiates rollback if boot_counter and satisfies
+fn trigger_rollback() -> Result<()> {
+    match handle_rollback() {
+        Ok(()) => {
+            log::info!("Rollback successful");
+            unset_boot_counter()?;
+            handle_reboot(true)
+        }
+        Err(e) => {
+            bail!("{e}, Rollback is not initiated");
+        }
+    }
 }
 
-fn stamp() -> Result<(), Error> {
-    fs::create_dir_all("/etc/greenboot/")?;
-    File::create("/etc/greenboot/upgrade.stamp")?;
-    Ok(())
+/// takes in a path and runs all the .sh files within the path
+/// returns false if any script fails
+fn run_scripts(name: &str, path: &str) -> Result<(), Vec<Box<dyn Error>>> {
```

Comment:
I think that this function should not return a Result, just a vector with the errors, the function itself never fails and thereby we don't propagate errors. What we are doing is collecting the errors of the commands, that could be 0 or many, hence the vector.